panther_hardware_interfaces/include/panther_hardware_interfaces/gpio_controller.hpp

```diff
@@ -0,0 +1,294 @@
+// Copyright 2023 Husarion sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file gpio_controller.hpp
+ * @brief Header file containing a higher-level wrapper for the panther_gpiod GPIO driver.
+ */
+
+#ifndef PANTHER_HARDWARE_INTERFACES_GPIO_CONTROLLER_HPP_
+#define PANTHER_HARDWARE_INTERFACES_GPIO_CONTROLLER_HPP_
+
+#include <atomic>
+#include <functional>
+#include <memory>
+#include <thread>
+#include <vector>
+
+#include <gpiod.hpp>
+
+#include "panther_gpiod/gpio_driver.hpp"
+
+namespace panther_hardware_interfaces
+{
+
+class Watchdog
+{
+public:
+  /**
+   * @brief Constructor for Watchdog class.
+   *
+   * @param gpio_driver Pointer to the GPIODriver object.
+   * @throws std::runtime_error if the Watchdog pin is not configured by GPIODriver or not described
+   * in GPIOController gpio_info storage
+   */
+  Watchdog(std::shared_ptr<panther_gpiod::GPIODriver> gpio_driver);
+
+  /**
+   * @brief Destructor for Watchdog class. Turns off the watchdog thread.
+   */
+  ~Watchdog();
+
+  /**
+   * @brief Turns on the Watchdog.
+   *
+   * @return True if successful; otherwise, false.
+   * @throws std::runtime_error if the Watchdog pin is not configured by GPIODriver or not described
+   * in GPIOController gpio_info storage
+   */
+  bool TurnOn();
+  bool TurnOff();
+  bool IsWatchdogEnabled() const;
+
+private:
+  /**
+   * @brief Monitors the Watchdog thread status.
+   *
+   * While the Watchdog thread is enabled, it toggles the state of the Watchdog pin between active
+   * and inactive states at regular intervals. The Watchdog thread is active for a duration of 10
+   * milliseconds and then sleeps for the same duration. When the Watchdog thread is disabled, it
+   * sets the Watchdog pin to an inactive state.
+   */
+  void WatchdogThread();
+
+  panther_gpiod::GPIOPin watchdog_pin_ = panther_gpiod::GPIOPin::WATCHDOG;
+  std::shared_ptr<panther_gpiod::GPIODriver> gpio_driver_;
+  std::unique_ptr<std::thread> watchdog_thread_;
+  std::atomic_bool watchdog_thread_enabled_ = false;
+};
+
+class GPIOControllerInterface
+{
+public:
+  virtual void Start() = 0;
```

Comment:
```suggestion
class GPIOControllerInterface
{
public:
  virtual ~GPIOControllerInterface() = default;
  virtual void Start() = 0;
```

Something like this. Should be tested before commit.